records_mover/records/hints.py

```diff
@@ -119,67 +120,99 @@ def sniff_hints_from_fileobjs(fileobjs: List[IO[bytes]],
-        logger.warning("Could not use chardet to detect encoding, as stream is not rewindable")
-        return None
+        raise OSError('Stream is not rewindable')
     original_position = fileobj.tell()
     fileobj.seek(0)
-    detector = chardet.UniversalDetector()
-    while True:
-        chunksize = 512
-        chunk = fileobj.read(chunksize)
-        detector.feed(chunk)
-        if detector.done or len(chunk) < chunksize:
-            break
-    detector.close()
-    fileobj.seek(original_position)
-    assert detector.result is not None
-    if 'encoding' in detector.result:
-        chardet_encoding = detector.result['encoding']
-        if chardet_encoding in hint_encoding_from_chardet:
-            return hint_encoding_from_chardet[chardet_encoding]
-        else:
-            logger.warning(f"Got unrecognized encoding from chardet sniffing: {detector.result}")
-            return None
-    else:
-        logger.warning(f"Unable to sniff file encoding using chardet: {detector.result}")
+    try:
+        yield fileobj
+    finally:
+        fileobj.seek(original_position)
+
+
+def sniff_encoding_hint(fileobj: IO[bytes]) -> Optional[HintEncoding]:
+    try:
+        with rewound_fileobj(fileobj) as fileobj:
+            detector = chardet.UniversalDetector()
+            while True:
+                chunksize = 512
+                chunk = fileobj.read(chunksize)
+                detector.feed(chunk)
+                if detector.done or len(chunk) < chunksize:
+                    break
+            detector.close()
+            assert detector.result is not None
+            if 'encoding' in detector.result:
+                chardet_encoding = detector.result['encoding']
+                if chardet_encoding in hint_encoding_from_chardet:
+                    return hint_encoding_from_chardet[chardet_encoding]
+                else:
+                    logger.warning("Got unrecognized encoding from chardet "
+                                   f"sniffing: {detector.result}")
+                    return None
+            else:
+                logger.warning(f"Unable to sniff file encoding using chardet: {detector.result}")
+                return None
+    except OSError:
+        logger.warning("Could not use chardet to detect encoding, as stream is not rewindable")
         return None
 
 
+def csv_hints_from_python(fileobj: IO[bytes],
+                          record_terminator_hint: HintRecordTerminator,
+                          encoding_hint: HintEncoding) -> RecordsHints:
+    # https://docs.python.org/3/library/csv.html#csv.Sniffer
+    try:
+        with rewound_fileobj(fileobj) as fileobj:
+            # Sniffer tries to determine quotechar, doublequote,
+            # delimiter, skipinitialspace.  does not try to determine
+            # lineterminator.
+            # https://github.com/python/cpython/blob/master/Lib/csv.py#L165
+
+            # TODO: need a text stream here, meaning we know encoding and
```

Comment:
Uncompleted punchlist item detected--consider resolving or moving this to your issue tracker